### PR TITLE
Enable optional client authentication to support mutual TLS with CQL storage backend.

### LINKED ISSUE #1324: No configuration to connect to cassandra with two-way SSL
It should be possible to connect a janusgraph client to a cassandra cluster which is set up with `require_client_auth: true` . Currently, the janusgraph configuration supports setting `storage.cassandra.ssl.truststore` parameters, but not for keystore.

This was tested with janusgraph 0.2.0 and the `cassandra` backend, but after reviewing the code it appears to be the same with janusgraph 0.3.1 and the `cql` backend.

Is there some configuration I'm missing or would this be a new feature?

### PR DESCRIPTION
**Pull request by callajd**
Fixes #1324

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [x] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?
- [x] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-cql/src/test/resources/docker/update_config.py [hunk 1]
```diff
@@ -16,9 +16,12 @@
 
 import os,re
 
-keystore = '/etc/ssl/test.keystore'
+keystore = '/etc/ssl/node.keystore'
 keystore_password = 'cassandra'
+truststore = '/etc/ssl/node.truststore'
+truststore_password = 'cassandra'
```
Comment (FlorianHockmann) [inline janusgraph-cql/src/test/resources/docker/update_config.py]: Codacy Issue found: [Possible hardcoded password: 'cassandra'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201)

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- janusgraph-cql/src/test/java/org/janusgraph/JanusGraphCassandraContainer.java  1
         

Clones added
============
- janusgraph-cql/src/test/resources/docker/update_config.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201&bid=15469658)

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- janusgraph-cql/src/test/java/org/janusgraph/JanusGraphCassandraContainer.java  1
         

Clones added
============
- janusgraph-cql/src/test/resources/docker/update_config.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201&bid=15469658)

File: janusgraph-cql/src/test/resources/docker/update_config.py (hunk 1 above)
Comment (FlorianHockmann) [inline janusgraph-cql/src/test/resources/docker/update_config.py]: Codacy Issue found: [Possible hardcoded password: 'cassandra'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201)

Comment (callajd) [reply janusgraph-cql/src/test/resources/docker/update_config.py]: This hardcoded password issue already existed in `update_config.py` (line 20) - it seems like we want to configure an exception.  If that's correct, can someone point me in the right direction?

Comment (farodin91) [reply janusgraph-cql/src/test/resources/docker/update_config.py]: You don’t have to set an exception. The tool isn’t a hard border for a pr to get merged.

File: docs/basics/janusgraph-cfg.md [hunk 2]
```diff
@@ -478,8 +478,19 @@ Configuration options for SSL
...
 
+### storage.cql.ssl.keystore
+Configuration options for SSL Keystore.
+
+
+| Name | Description | Datatype | Default Value | Mutability |
+| ---- | ---- | ---- | ---- | ---- |
+| storage.cql.ssl.keystore.keypassword | The password to access the key in SSL Keystore. | String |  | LOCAL |
+| storage.cql.ssl.keystore.location | Marks the location of the SSL Keystore. | String |  | LOCAL |
+| storage.cql.ssl.keystore.storepassword | The password to access the store in SSL Keystore. | String |  | LOCAL |
```
Comment (callajd) [inline docs/basics/janusgraph-cfg.md]: Will change this to "The password to access the SSL Keystore."

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- janusgraph-cql/src/test/java/org/janusgraph/JanusGraphCassandraContainer.java  1
         

Clones added
============
- janusgraph-cql/src/test/resources/docker/update_config.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201&bid=15469658)

**Comment by FlorianHockmann** [comment]
One last thing from my side that I completely forgot about: Could you please also [squash your commits](https://github.com/wprig/wprig/wiki/How-to-squash-commits) as our git workflow is to use only one commit per contribution?

File: janusgraph-cql/src/test/resources/docker/update_config.py (hunk 1 above)
Comment (FlorianHockmann) [inline janusgraph-cql/src/test/resources/docker/update_config.py]: Codacy Issue found: [Possible hardcoded password: 'cassandra'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201)

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- janusgraph-cql/src/test/java/org/janusgraph/JanusGraphCassandraContainer.java  1
         

Clones added
============
- janusgraph-cql/src/test/resources/docker/update_config.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201&bid=15469658)

**Comment by porunov** [comment]
@callajd Please, squash your commits into a single commit and rebase against current `master` branch (because there are conflicts in `.travis.yml` which prevent us from merging this PR)

**Comment by porunov** [comment]
@callajd We are planning to release `0.5.0` version soon. Will you be able to resolve above comments so that we could merge this PR?
If you are busy right now, we can just leave this feature for future releases

**Comment by callajd** [comment]
Hi, I should be able to address this over the weekend - does that work for release schedule?

**Comment by porunov** [comment]
@callajd Yes, for sure. We can wait that time. Thank you

**Comment by callajd** [comment]
My apologies, I was pulled away from this over the weekend.  I will wrap this up tonight.

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- janusgraph-cql/src/test/resources/docker/update_config.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201&bid=15469658)

File: janusgraph-cql/src/test/resources/docker/update_config.py (hunk 1 above)
Comment (FlorianHockmann) [inline janusgraph-cql/src/test/resources/docker/update_config.py]: Codacy Issue found: [Possible hardcoded password: 'cassandra'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201)

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- janusgraph-cql/src/test/resources/docker/update_config.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201&bid=15469658)

File: janusgraph-cql/src/test/resources/docker/update_config.py (hunk 1 above)
Comment (FlorianHockmann) [inline janusgraph-cql/src/test/resources/docker/update_config.py]: Codacy Issue found: [Possible hardcoded password: 'cassandra'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201)

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- janusgraph-cql/src/test/resources/docker/update_config.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201&bid=15469658)

File: janusgraph-cql/src/test/resources/docker/update_config.py (hunk 1 above)
Comment (FlorianHockmann) [inline janusgraph-cql/src/test/resources/docker/update_config.py]: Codacy Issue found: [Possible hardcoded password: 'cassandra'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4611201)

**Comment by porunov** [comment]
@callajd Tell if you need any help in squashing your commits into a single one. Currently your PR includes 10 commits (4 your commits and 6 commits from others).
See https://github.com/JanusGraph/janusgraph/pull/1900/commits
We need to make sure that this PR contains only 1 commit. No any other commits

**Comment by callajd** [comment]
@porunov I understand; I used by day job's workflow out of habit so I'm trying to clean things up now.

**Comment by porunov** [comment]
@callajd You can fix those problems with history by applying the next commands in your branch:
```
git remote add upstream https://github.com/JanusGraph/janusgraph
git fetch --all
git reset --hard upstream/master
git merge --squash origin/cql-tls-support
git add -A
git commit -m "Enable optional client authentication to support mutual TLS with CQL storage backend [full build]" -m "Fixes #1324" -s
git push -f
```

Or you can just fetch the [branch](https://github.com/porunov/janusgraph/tree/cql-tls-support) from my repository, reset your branch against my branch, change the author of the commit and push it as is:
```
git remote add porunov https://github.com/porunov/janusgraph
git fetch --all
git reset --hard porunov/cql-tls-support
git reset HEAD~1 --soft
git commit -m "Enable optional client authentication to support mutual TLS with CQL storage backend [full build]" -m "Fixes #1324" -s
git push -f
```

**Comment by porunov** [comment]
@callajd Are you still interested in resolving this PR? If you can't work on it anymore I would like to provide this feature based on your contribution

**Comment by porunov** [comment]
@callajd I have created a separate PR which is a plain copy paste of your contribution here #1988 
If you are able to resolve the history issues with your PR, I would ask you to do it before we merge #1988 so that we could merge this PR instead of #1988

**Comment by porunov** [comment]
The squashed commit was merged in this PR: #1988 
Thank you @callajd for the contribution!

**Pull request closed without merging**